src/core/node.ts

```diff
@@ -151,15 +151,24 @@ export class Node {
     }
 
     prepareChild(subpath: string, child: any): any {
-        const childFactory = this.getChildFactory(subpath)
-        invariant(childFactory.is(child), `The given value is not assignable to "${subpath}": ${child}`)
         if (!isMutable(child)) {
             return child
         }
+        
+        const childFactory = this.getChildFactory(subpath)
 
         if (hasNode(child)) {
+            const node = getNode(child)
+
+            // TODO: reason about this... Mh... :/
+            // we are adding a node with no parent (first insert in the tree)
+            if(node.parent === null){
+                node.setParent(this, subpath)
+                return child
+            }
+
             // TODO: fail here or implicitly clone?
-            child = getNode(child).snapshot
+            child = node.snapshot
```

Comment:
I think we should throw here, implicit cloning might be to subtle, especially since we don't clone anymore for complete trees